crates/compiler/builtins/roc/Json.roc

```diff
@@ -254,230 +255,184 @@ encodeTag = \name, payload ->
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toU8
+        |> Result.mapErr \_ -> TooShort
+
+    { result, rest }
 
 decodeU16 = Decode.custom \bytes, @Json {} ->
-    { taken, rest } = takeDigits bytes
+    { taken, rest } = takeJsonNumber bytes
+
+    result =
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toU16
+        |> Result.mapErr \_ -> TooShort
 
-    when Str.fromUtf8 taken |> Result.try Str.toU16 is
-        Ok n -> { result: Ok n, rest }
-        Err _ -> { result: Err TooShort, rest }
+    { result, rest }
 
 decodeU32 = Decode.custom \bytes, @Json {} ->
-    { taken, rest } = takeDigits bytes
+    { taken, rest } = takeJsonNumber bytes
 
-    when Str.fromUtf8 taken |> Result.try Str.toU32 is
-        Ok n -> { result: Ok n, rest }
-        Err _ -> { result: Err TooShort, rest }
+    result =
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toU32
+        |> Result.mapErr \_ -> TooShort
+
+    { result, rest }
 
 decodeU64 = Decode.custom \bytes, @Json {} ->
-    { taken, rest } = takeDigits bytes
+    { taken, rest } = takeJsonNumber bytes
+
+    result =
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toU64
+        |> Result.mapErr \_ -> TooShort
 
-    when Str.fromUtf8 taken |> Result.try Str.toU64 is
-        Ok n -> { result: Ok n, rest }
-        Err _ -> { result: Err TooShort, rest }
+    { result, rest }
 
 decodeU128 = Decode.custom \bytes, @Json {} ->
-    { taken, rest } = takeDigits bytes
+    { taken, rest } = takeJsonNumber bytes
+
+    result =
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toU128
+        |> Result.mapErr \_ -> TooShort
 
-    when Str.fromUtf8 taken |> Result.try Str.toU128 is
-        Ok n -> { result: Ok n, rest }
-        Err _ -> { result: Err TooShort, rest }
+    { result, rest }
 
 decodeI8 = Decode.custom \bytes, @Json {} ->
-    { taken, rest } = takeDigits bytes
+    { taken, rest } = takeJsonNumber bytes
 
-    when Str.fromUtf8 taken |> Result.try Str.toI8 is
-        Ok n -> { result: Ok n, rest }
-        Err _ -> { result: Err TooShort, rest }
+    result =
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toI8
+        |> Result.mapErr \_ -> TooShort
+
+    { result, rest }
 
 decodeI16 = Decode.custom \bytes, @Json {} ->
-    { taken, rest } = takeDigits bytes
+    { taken, rest } = takeJsonNumber bytes
+
+    result =
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toI16
+        |> Result.mapErr \_ -> TooShort
 
-    when Str.fromUtf8 taken |> Result.try Str.toI16 is
-        Ok n -> { result: Ok n, rest }
-        Err _ -> { result: Err TooShort, rest }
+    { result, rest }
 
 decodeI32 = Decode.custom \bytes, @Json {} ->
-    { taken, rest } = takeDigits bytes
+    { taken, rest } = takeJsonNumber bytes
+
+    result =
+        taken
+        |> Str.fromUtf8
+        |> Result.try Str.toI32
+        |> Result.mapErr \_ -> TooShort
```

Comment:
🎨 FYI, this can also be written as:

```elm
        |> Result.mapErr \InvalidNumStr -> TooShort
```

I've been trying this style out (using the tag name instead of `_` when there's only a single tag) and liking the self-documentation aspect so far.

Reply:
@rtfeldman I gave it a go but the type is actually `[BadUtf8 Utf8ByteProblem Nat, InvalidNumStr]*` so maybe I'll leave it for now as `_` is more concise. I have been thinking, should we have more than just a `TooShort` option for `DecodeError` to be more descriptive? Almost all decoders will be doing some kind of utf-8 parsing and then conversion to a RocType. Perhaps adding a `BadUtf8`, `UnexpectedType` and `InvalidType` might be useful?

Reply:
Also posted to zulip [discussion](https://roc.zulipchat.com/#narrow/stream/304641-ideas/topic/Decode.20Errors/near/344787617)